src/hotspot/share/oops/markWord.hpp

```diff
@@ -246,8 +246,7 @@ class markWord {
 
   // Should this header be preserved during GC?
   bool must_be_preserved() const {
-    return (!is_unlocked() || !has_no_hash() ||
-      (EnableValhalla && (is_larval_state() || is_inline_type() || is_flat_array() || is_null_free_array())));
+    return (!is_unlocked() || !has_no_hash() || (EnableValhalla && is_larval_state()));
```

Comment:
FYI: the `EnableValhalla` is not necessary and I am removing it with my changes. I guess whoever integrates last can make sure it's gone.